compiler/src/iree/compiler/Dialect/HAL/Transforms/ResolveTopologyQueries.cpp

```diff
@@ -0,0 +1,191 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <utility>
+
+#include "iree/compiler/Dialect/HAL/Analysis/DeviceAnalysis.h"
+#include "iree/compiler/Dialect/HAL/Conversion/StreamToHAL/Utils.h"
+#include "iree/compiler/Dialect/HAL/IR/HALDialect.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "iree/compiler/Dialect/HAL/Transforms/Passes.h"
+#include "iree/compiler/Dialect/Stream/IR/StreamDialect.h"
+#include "iree/compiler/Dialect/Stream/IR/StreamOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-hal-resolve-topology-queries"
+
+namespace mlir::iree_compiler::IREE::HAL {
+
+#define GEN_PASS_DEF_RESOLVETOPOLOGYQUERIESPASS
+#include "iree/compiler/Dialect/HAL/Transforms/Passes.h.inc"
+
+namespace {
+
+// Check if all affinities of the optimal attribute refer to the same device ID
+static bool allReferToSameDevice(DeviceOptimalAttr optimalAttr,
+                                 DeviceAnalysis &deviceAnalysis,
+                                 Operation *fromOp) {
+  SetVector<IREE::HAL::DeviceTargetAttr> allPossibleTargets;
+  for (auto affinity : optimalAttr.getAffinities()) {
+    deviceAnalysis.gatherDeviceAffinityTargets(affinity, fromOp,
+                                               allPossibleTargets);
+  }
+
```

Comment:
NIT: No need to create another set vector

```suggestion
if (allPossibleTargets.empty()) {
  return false;
}

if (allPossibleTargets.size() == 1) {
   return true;
}

const auto firstDeviceId = allPossibleTargets.at(0);
for (const auto &target : allPossibleTargets) {
    if (firstDeviceId != target.getDeviceID()) {
         return false;
    }
}
```